Add infrastructure for generic input backends to feed devices

sdiazpier opened the pull request:
This PR adds the infrastructure required for the input part of NESTio.

---
See #1283 for the RecordingBackend counterpart of the NESTio infrastructure.

hakonsbm commented on nestkernel/input_backend_internal.cpp:
```diff
@@ -0,0 +1,143 @@
+  thread tid = device.get_thread();
+  index node_id = device.get_node_id();
+
+  device_map::value_type::iterator device_it = devices_[ tid ].find( node_id );
```
```suggestion
  auto device_it = devices_[ tid ].find( node_id );
```

sdiazpier replied:
done

hakonsbm commented on nestkernel/input_backend_mpi.cpp:
```diff
@@ -0,0 +1,299 @@
+    thread tid = device.get_thread();
+    index node_id = device.get_node_id();
+
+    device_map::value_type::iterator device_it = devices_[ tid ].find( node_id );
```
```suggestion
    auto device_it = devices_[ tid ].find( node_id );
```

sdiazpier replied:
done

hakonsbm commented on nestkernel/input_backend_mpi.cpp:
```diff
@@ -0,0 +1,299 @@
+  thread tid = device.get_thread();
+  index node_id = device.get_node_id();
+
+  device_map::value_type::iterator device_it = devices_[ tid ].find( node_id );
```
```suggestion
  auto device_it = devices_[ tid ].find( node_id );
```

sdiazpier replied:
done

hakonsbm commented on nestkernel/input_backend_mpi.cpp:
```diff
@@ -0,0 +1,299 @@
+    // add the link between MPI communicator and the device (device can share the same MPI communicator
+    std::string port_name;
+    get_port( it_device->second.second, &port_name );
+    comm_map::value_type::iterator comm_it = commMap_[ thread_id ].find( port_name );
```
```suggestion
    auto comm_it = commMap_[ thread_id ].find( port_name );
```

sdiazpier replied:
done

hakonsbm commented on nestkernel/input_backend_mpi.cpp:
```diff
@@ -0,0 +1,299 @@
+  // 2) connect the thread with MPI process it need to be connected
+  // WARNING can be a bug if it's need all the thread to be connected in MPI
+  comm_map::value_type::iterator it_comm;
+  for ( it_comm = commMap_[ thread_id ].begin(); it_comm != commMap_[ thread_id ].end(); it_comm++ )
```
Use `auto`, or a range-based for loop if possible.

sdiazpier replied:
done

hakonsbm commented on nestkernel/input_backend_mpi.cpp:
```diff
@@ -0,0 +1,299 @@
+  // connect take information of MPI process (for the moment only spike train)
+  const thread thread_id = kernel().vp_manager.get_thread_id();
+  device_map::value_type::iterator it_device;
+  for ( it_device = devices_[ thread_id ].begin(); it_device != devices_[ thread_id ].end(); it_device++ )
```
Use `auto`, or a range-based for loop if possible.

sdiazpier replied:
done

hakonsbm commented on models/spike_generator.h:
```diff
@@ -212,7 +213,9 @@ class spike_generator : public DeviceNode
   port send_test_event( Node&, rport, synindex, bool );
   void get_status( DictionaryDatum& ) const;
   void set_status( const DictionaryDatum& );
+  void update_from_backend( std::vector< double > input_spikes ) override;
```
Marking only this function `override` is inconsistent. If this is marked `override`, then the following functions should also be marked `override`:
- `has_proxies`
- `get_element_type`
- `send_test_event`
- `get_status`
- `set_status`
- `get_type`
- `event_hook`
- `sends_signal`
- `init_state_`
- `init_buffers_`
- `calibrate`
- `update`

lionelkusch replied:
Do you prefer to indicate all the function override or not?

hakonsbm replied:
Marking all functions listed `override` is preferred, because it better conveys the intent and can help catch bugs at compile time.

lionelkusch replied:
done

hakonsbm commented on nestkernel/input_device.h:
```diff
@@ -0,0 +1,122 @@
+  InputDevice();
+  InputDevice( const InputDevice& );
+
+  void calibrate( const std::vector< Name >&, const std::vector< Name >& );
```
This hides overloaded virtual functions. You can add `using` directives to avoid hiding overloaded virtual functions:
```c++
using Device::calibrate;
```
See also #1377.

sdiazpier replied:
done

hakonsbm commented on nestkernel/input_device.h:
```diff
@@ -0,0 +1,122 @@
+
+protected:
+  std::vector< double > read();
+  void set_initialized_() override;
```
Marking only this function `override` is inconsistent. If this is marked `override`, then the following functions should also be marked `override`:
- `is_active`
- `set_status`
- `get_status`

sdiazpier replied:
done

sdiazpier commented:
Thank you dear @hakonsbm we will work on these corrections right away.

lionelkusch commented:
Thank you @hakonsbm for all requests. @sdiazpier  and me fix all of this. 
If you have new requests don't hesitate to ask them.

hakonsbm commented on nestkernel/io_manager.cpp:
```diff
@@ -178,6 +198,19 @@ IOManager::set_status( const DictionaryDatum& d )
+  DictionaryDatum input_backends;
+  if ( updateValue< DictionaryDatum >( d, names::input_backends, input_backends ) )
+  {
+    std::map< Name, InputBackend* >::const_iterator iti;
+    for ( iti = input_backends_.begin(); iti != input_backends_.end(); ++iti )
```
I think you can use a range-based for loop here.

sdiazpier replied:
done

hakonsbm commented on nestkernel/recording_backend_mpi.cpp:
```diff
@@ -0,0 +1,300 @@
+nest::RecordingBackendMPI::prepare()
+{
+  // Create the connection with MPI per thread
+  // TODO validate approach
```
Should this TODO be fixed?

sdiazpier replied:
Dear @hakonsbm this TODO is a bit complicated to solve. We have asked @jougs for a meeting to discuss this issue.

sdiazpier replied:
solved

hakonsbm commented on nestkernel/input_backend.h:
```diff
@@ -0,0 +1,280 @@
+
+} // namespace
+
+#endif // RECORDING_BACKEND_H
```
This should be corrected.

sdiazpier replied:
done

hakonsbm commented on nestkernel/io_manager.h:
```diff
@@ -91,26 +92,38 @@ class IOManager : public ManagerInterface
-  void cleanup();
-  void prepare();
+  void cleanup() override;
+  void prepare() override;
```
For consistency the following functions in this class should also be marked `override`:
- `initialize`
- `finalize`
- `change_num_threads`
- `set_status`
- `get_status`

sdiazpier replied:
done

hakonsbm commented on nestkernel/input_backend_mpi.cpp:
```diff
@@ -0,0 +1,295 @@
+  int shape[ 1 ];
+  MPI_Recv( &shape, 1, MPI_INT, MPI_ANY_SOURCE, message[ 1 ], comm, &status_mpi );
+  // Receive the data ( for the moment only spike time )
+  double spikes[ shape[ 0 ] ];
```
My compiler gives a warning that using variable length arrays is a C99 feature. As we have switched to C++11, we should also be using C11, I think. However, this is not specified when compiling. I can create an issue on this.

lionelkusch replied:
I choice variable length array because it's impossible to know the number of spike receiving before.
For avoid this, I need to modify the protocol of communication in order to define a fix number of spike receiving. This modifiication will complex the communication and can slow it but it will increase the robustness of the code.
What do you prefer?

hakonsbm replied:
I would prefer avoiding variable length arrays. And I see that you have already changed to dynamic allocation, which should do the trick.

jougs commented on models/spike_generator.cpp:
```diff
@@ -168,15 +167,29 @@ nest::spike_generator::Parameters_::assert_valid_spike_time_and_insert_( double
   bool flags_changed = updateValueParam< bool >( d, names::precise_times, precise_times_, node )
     or updateValueParam< bool >( d, names::shift_now_spikes, shift_now_spikes_, node )
     or updateValueParam< bool >( d, names::allow_offgrid_times, allow_offgrid_times_, node );
+
+  if ( d->known( names::allow_offgrid_times ) )
+  {
+    allow_offgrid_times_ = getValue< bool >( d->lookup( names::allow_offgrid_times ) );
+  }
+  if ( d->known( names::precise_times ) )
+  {
+    precise_times_ = getValue< bool >( d->lookup( names::precise_times ) );
+  }
+  if ( d->known( names::shift_now_spikes ) )
+  {
+    shift_now_spikes_ = getValue< bool >( d->lookup( names::shift_now_spikes ) );
+  }
+
```
Why is this here? If I'm not completely mistaken, the block above already updates the values if they are known. This is what the `updateValueParam()` function is supposed to do. @hakonsbm, please correct me if I'm wrong.

hakonsbm replied:
That's right, `updateValueParam()` will update the variable passed to it if the name is known, and return a `bool` stating if the variable was updated. There is therefore no need for these `if`-tests.

sdiazpier replied:
solved

lionelkusch commented:
cc @maedoc

jougs commented:
@lionelkusch, @sdiazpier: can you please give a short update of the status? Many thanks!

jougs commented:
This is just a friendly ping!

sdiazpier commented:
Dear @jougs we are still working on details merging the stimulating device with the input device. However, the more simple things we have mostly addressed already:

- Making StimulatingDevice a base class of the concrete stimulator -> Done
- To further simplify the design, the functionality of the new base class InputDevice should be included into the StimulatingDevice class.-> Problematic, in progress

- The base class for recording backends was first called OutputBackend. We decided to rename them to RecordingBackends to make clear that they are not backends for general output operations from within NEST, but only handle the output generated by recorders. The same considerations apply to the naming of the current InputBackend base class, which IMO should be renamed to StimulatingBackend -> Done

- Refactoring of SpikeParameters_ struct and State_ struct. -> Done

- Why is the InputBackendInternal needed at all if it does nothing? Couldn't an empty input_from field just be interpreted in this way? -> In progress

- The InputBackend base class is currently more or less a copy of the RecordingBackend, which raises a number of questions:
  * The documentation is still mostly describing the design and workings of the recording backend infrastructure with just the word "recording" replaced by "input". -> Done
  * I don't understand how some of the interface functions that were taken from the recording backend design are actually useful for the input side. -> Removed set_value_names() and unnecessary hooks.

- Renaming:
* input_from ⟶ stimulus_source or data_source -> Done to stimulus_source
* MPIFilePortsUnknown ⟶ MPIPortsFileUnknown -> Done
* Change name of function update_from_backend() to set_data_from_input_backend() -> Done to set_data_from_stimulating_backend()

As soon as we finalize the open points we will update the PR.

jougs commented:
Awesome! Many thanks for the update and all the work in the background :-)

lionelkusch commented:
I get an issue due to the use of different implementations of MPI. 

I am using mpich-3.1.4 for MPI. I did have any problems for compiling Nest.
However, Travis and @sdiazpier use openmpi. The compiler complains because the variable ompi_mpi_info_null is not shared by the threads in the file nest-simulator/nestkernel/recording_backend_mpi.cpp, line 175. If I add them, I can't compile because the variable doesn't exist. 

Do you know how to solve this problem?
Do we need to write a code aware of the MPI implementation?

lionelkusch commented:
I fix the 

> Do you know how to solve this problem?
> Do we need to write a code aware of the MPI implementation?

I fix the problem by changing the parallelization of the recorder.
We need to note :  the shared variable between thread when we are using MPI command in openMP parallel depend on the MPI implementation.

jougs reviewed (commented):
Many thanks for your updates!

I have added some concrete remarks as inline comments. In addition to looking at those, can you please make sure that all comments from my last review in April 2020 were addressed? I found that some of those points are still standing (e.g. `set_value_name()` is still there and the documentation is not fully updated).

Would it make sense to have a common `IOBackend` baseclass for stimulating and recording backends? That way we could have a single registry for them in `IOManager` and things like initialization, parameter getting/setting could become a bit simpler and we would have to duplicate less code. This does not have to happen right away, I'd rather like to hear your thoughts on this.

lionelkusch commented on nestkernel/exceptions.h:
```diff
@@ -1180,6 +1180,23 @@ class MUSICChannelAlreadyMapped : public KernelException
+class MPIFilePortsUnknown : public KernelException
+{
+public:
+  explicit MPIFilePortsUnknown( const index index_node )
```
done

lionelkusch commented on nestkernel/input_backend_mpi.h:
```diff
@@ -0,0 +1,133 @@
+Communication Protocol:
++++++++++++++++++++++++
+
+To get the information to update the device, MPI backend sends the GID and the id
```
done

lionelkusch commented on nestkernel/exceptions.cpp:
```diff
@@ -387,6 +387,18 @@ nest::MUSICChannelAlreadyMapped::message() const
+  std::ostringstream msg;
+  msg << "For the node " << index_node_ << ", the label "
+      << "needs to be defined. The label precises "
+      << "the folder of files for the MPI port";
```
done

lionelkusch commented on nestkernel/exceptions.h:
```diff
@@ -1180,6 +1180,23 @@ class MUSICChannelAlreadyMapped : public KernelException
+{
+public:
+  explicit MPIFilePortsUnknown( const index index_node )
+    : index_node_( index_node )
```
done

lionelkusch commented on nestkernel/exceptions.h:
```diff
@@ -1180,6 +1180,23 @@ class MUSICChannelAlreadyMapped : public KernelException
+  std::string message() const;
+
+private:
+  const index index_node_;
```
done

lionelkusch commented on nestkernel/exceptions.h:
```diff
@@ -1180,6 +1180,23 @@ class MUSICChannelAlreadyMapped : public KernelException
+private:
+  const index index_node_;
+};
+
```
done

lionelkusch commented on models/ac_generator.cpp:
```diff
@@ -236,3 +234,34 @@ nest::ac_generator::handle( DataLoggingRequest& e )
+  {
+    if ( input_param.size() != 4 )
+    {
+      throw BadParameterValue( "The size of the data for the ac_generator is incorrect." );
```
miss documentation

sdiazpier replied:
Added documentation to ac_generator.h in a new section called Update from stimulating backend. Please check if this is ok.

lionelkusch commented on models/ac_generator.cpp:
```diff
@@ -236,3 +234,34 @@ nest::ac_generator::handle( DataLoggingRequest& e )
+
+  // if we get here, temporary contains consistent set of properties
+  P_ = ptmp;
+}
```
done

lionelkusch commented on models/ac_generator.h:
```diff
@@ -189,9 +176,8 @@ class ac_generator : public DeviceNode
-    double phi_rad_; //!< Phase of sine current (0-2Pi rad)
-
+    // double omega_;   //!< Angelfrequency i rad/s
+    // double phi_rad_; //!< Phase of sine current (0-2Pi rad)
```
done

lionelkusch commented on models/dc_generator.cpp:
```diff
@@ -192,3 +189,27 @@ nest::dc_generator::handle( DataLoggingRequest& e )
+  {
+    if ( input_param.size() != 1 )
+    {
+      throw BadParameterValue( "The size of the data for the dc_generator is incorrect." );
```
miss documentation

sdiazpier replied:
Updated doc in .h file

lionelkusch commented on models/dc_generator.h:
```diff
@@ -149,7 +137,7 @@ class dc_generator : public DeviceNode
     State_(); //!< Sets default parameter values
 
-    void get( DictionaryDatum& ) const; //!< Store current values in dictionary
+    void get( DictionaryDatum& ) const {}; //!< Store current values in dictionary
```
done

lionelkusch commented on models/gamma_sup_generator.cpp:
```diff
@@ -280,3 +278,33 @@ nest::gamma_sup_generator::event_hook( DSSpikeEvent& e )
+
+  // if we get here, temporary contains consistent set of properties
+  P_ = ptmp;
+}
```
done

lionelkusch commented on models/gamma_sup_generator.cpp:
```diff
@@ -280,3 +278,33 @@ nest::gamma_sup_generator::event_hook( DSSpikeEvent& e )
+  {
+    if ( input_param.size() != 4 )
+    {
+      throw BadParameterValue( "The size of the data for the ac_generator is incorrect." );
```
miss documentation

sdiazpier replied:
Updated doc in .h file

lionelkusch commented on models/mip_generator.cpp:
```diff
@@ -218,3 +216,32 @@ nest::mip_generator::event_hook( DSSpikeEvent& e )
+  {
+    if ( input_param.size() != 2 )
+    {
+      throw BadParameterValue( "The size of the data for the ac_generator is incorrect." );
```
miss documention

sdiazpier replied:
Updated doc in .h file

lionelkusch commented on models/noise_generator.cpp:
```diff
@@ -361,3 +359,37 @@ nest::noise_generator::handle( DataLoggingRequest& e )
+  {
+    if ( input_param.size() != 4 )
+    {
+      throw BadParameterValue( "The size of the data for the ac_generator is incorrect." );
```
miss documentation

sdiazpier replied:
updated doc in .h

lionelkusch commented on models/noise_generator.h:
```diff
@@ -262,9 +250,9 @@ class noise_generator : public DeviceNode
-    double phi_rad_;                        //!< Phase of sine current (0-2Pi rad)
 
+    // double omega_;                          //!< Angelfrequency i rad/s
+    // double phi_rad_;                        //!< Phase of sine current (0-2Pi rad)
```
done

lionelkusch commented on models/poisson_generator.cpp:
```diff
@@ -153,3 +151,31 @@ nest::poisson_generator::event_hook( DSSpikeEvent& e )
+  {
+    if ( input_param.size() != 1 )
+    {
+      throw BadParameterValue( "The size of the data for the poisson generator is incorrect." );
```
miss documentation

sdiazpier replied:
updated doc in .h file

lionelkusch commented on models/ac_generator.cpp:
```diff
@@ -236,3 +234,34 @@ nest::ac_generator::handle( DataLoggingRequest& e )
+      ( *d )[ names::frequency ] = DoubleDatum( input_param[ 2 ] );
+      ( *d )[ names::phase ] = DoubleDatum( input_param[ 3 ] );
+      ptmp.set( d, this );
+    }
```
done for all

lionelkusch commented on models/poisson_generator_ps.cpp:
```diff
@@ -284,5 +280,29 @@ nest::poisson_generator_ps::event_hook( DSSpikeEvent& e )
+  {
+    if ( input_param.size() != 2 )
+    {
+      throw BadParameterValue( "The size of the data for the poisson_generator_ps is incorrect." );
```
miss documentation

sdiazpier replied:
updated in .h file

lionelkusch commented on models/ppd_sup_generator.cpp:
```diff
@@ -292,3 +287,35 @@ nest::ppd_sup_generator::event_hook( DSSpikeEvent& e )
+  {
+    if ( input_param.size() != 5 )
+    {
+      throw BadParameterValue( "The size of the data for the ppd_sup_generator is incorrect." );
```
miss documentation

sdiazpier replied:
Updated doc in .h file

lionelkusch commented on models/pulsepacket_generator.cpp:
```diff
@@ -237,3 +235,34 @@ nest::pulsepacket_generator::update( Time const& T, const long from, const long
+  {
+    if ( input_param.size() < 3 )
+    {
+      throw BadParameterValue( "The size of the data for the  pulse_generator is incorrect." );
```
miss documentation

sdiazpier replied:
Updated doc in .h file

lionelkusch commented on models/step_current_generator.cpp:
```diff
@@ -328,3 +321,40 @@ nest::step_current_generator::handle( DataLoggingRequest& e )
+{
+  Parameters_ ptmp = P_; // temporary copy in case of errors
+
+  assert( time_amplitude.size() % 2 == 0 );
```
done

lionelkusch commented on models/inhomogeneous_poisson_generator.cpp:
```diff
@@ -291,3 +289,40 @@ nest::inhomogeneous_poisson_generator::event_hook( DSSpikeEvent& e )
+  // For the input backend
+  if ( not rate_time_update.empty() )
+  {
+
```
done

sdiazpier commented on models/spike_generator.cpp:
```diff
@@ -168,15 +170,32 @@ nest::spike_generator::Parameters_::assert_valid_spike_time_and_insert_( double
   Node* node )
 {
+
+
```
done

sdiazpier commented on models/spike_generator.cpp:
```diff
@@ -168,15 +170,32 @@ nest::spike_generator::Parameters_::assert_valid_spike_time_and_insert_( double
     or updateValueParam< bool >( d, names::shift_now_spikes, shift_now_spikes_, node )
     or updateValueParam< bool >( d, names::allow_offgrid_times, allow_offgrid_times_, node );
+
+
```
done

sdiazpier commented on models/spike_generator.cpp:
```diff
@@ -334,6 +354,45 @@ nest::spike_generator::calibrate()
+      times_ms.push_back( input_spike );
+      if ( ptmp.precise_times_ )
+      {
+        printf( "precise not supported\n" );
```
done

sdiazpier commented on models/spike_generator.h:
```diff
@@ -268,8 +263,11 @@ class spike_generator : public DeviceNode
-    Parameters_();                     //!< Sets default parameter values
-    Parameters_( const Parameters_& ); //!< Recalibrate all times
+    //!< Array of input backends to use
+    // Name input_from_;
```
done

sdiazpier commented on nestkernel/input_backend_mpi.cpp:
```diff
@@ -0,0 +1,299 @@
+// Includes from nestkernel:
+#include "input_backend_mpi.h"
+#include "input_device.h"
+#include <fstream>
```
done

sdiazpier commented on nestkernel/input_backend_mpi.cpp:
```diff
@@ -0,0 +1,299 @@
+{
+  // clear vector of map
+  device_map::iterator it_device;
+  for ( it_device = devices_.begin(); it_device != devices_.end(); it_device++ )
```
done

sdiazpier commented on nestkernel/input_backend_mpi.cpp:
```diff
@@ -0,0 +1,299 @@
+    it_device->clear();
+  }
+  comm_map::iterator it_comm;
+  for ( it_comm = commMap_.begin(); it_comm != commMap_.end(); it_comm++ )
```
done

sdiazpier commented on nestkernel/input_backend_mpi.cpp:
```diff
@@ -0,0 +1,299 @@
+
+  // get port and update the list of device
+  device_map::value_type::iterator it_device;
+  for ( it_device = devices_[ thread_id ].begin(); it_device != devices_[ thread_id ].end(); it_device++ )
```
done

sdiazpier commented on nestkernel/input_backend_mpi.cpp:
```diff
@@ -0,0 +1,299 @@
+      0,
+      MPI_COMM_WORLD,
+      it_comm->second.first ); // should use the status for handle error
+    printf( "Connect to %s\n", it_comm->first.data() );
```
done

sdiazpier commented on nestkernel/input_backend_mpi.cpp:
```diff
@@ -0,0 +1,299 @@
+  thread thread_id = kernel().vp_manager.get_thread_id();
+  // WARNING can be a bug if all the thread to send ending connection of MPI
+  comm_map::value_type::iterator it_comm;
+  for ( it_comm = commMap_[ thread_id ].begin(); it_comm != commMap_[ thread_id ].end(); it_comm++ )
```
done

sdiazpier commented on nestkernel/input_backend_mpi.cpp:
```diff
@@ -0,0 +1,299 @@
+  // WARNING can be a bug if all the thread to send ending connection of MPI
+  // disconnect MPI
+  comm_map::value_type::iterator it_comm;
+  for ( it_comm = commMap_[ thread_id ].begin(); it_comm != commMap_[ thread_id ].end(); it_comm++ )
```
done

sdiazpier commented on nestkernel/input_backend_mpi.cpp:
```diff
@@ -0,0 +1,299 @@
+  {
+    int value[ 1 ];
+    value[ 0 ] = thread_id;
+    MPI_Send( value, 1, MPI_INT, 0, 1, *it_comm->second.first );
```
done

sdiazpier commented on nestkernel/input_backend_mpi.cpp:
```diff
@@ -0,0 +1,299 @@
+  {
+    int value[ 1 ];
+    value[ 0 ] = thread_id;
+    MPI_Send( value, 1, MPI_INT, 0, 2, *it_comm->second.first );
```
done

sdiazpier commented on nestkernel/input_backend_mpi.cpp:
```diff
@@ -0,0 +1,299 @@
+nest::InputBackendMPI::get_port( const index index_node, const std::string& label, std::string* port_name )
+{
+  std::ostringstream
+    basename; // path of the file : path+label+id+.txt (file contains only one line with name of the port
```
done

sdiazpier commented on nestkernel/input_backend_mpi.cpp:
```diff
@@ -0,0 +1,299 @@
+  // clear map of device
+  commMap_[ thread_id ].clear();
+  device_map::value_type::iterator it_device;
+  for ( it_device = devices_[ thread_id ].begin(); it_device != devices_[ thread_id ].end(); it_device++ )
```
done

sdiazpier commented on nestkernel/input_backend_mpi.cpp:
```diff
@@ -0,0 +1,299 @@
+  }
+  else
+  {
+    // TODO take in count this case
```
done

sdiazpier commented on nestkernel/input_device.cpp:
```diff
@@ -0,0 +1,240 @@
+  updateValue< bool >( d, names::time_in_steps, time_in_steps );
+
+  // if ( time_in_steps != time_in_steps_ and S_.n_events != 0 ) //I don't understand n_events
+  // probabily link with S_.n_events_
```
done

sdiazpier commented on nestkernel/input_device.cpp:
```diff
@@ -0,0 +1,240 @@
+{
+  const long stamp = T.get_steps();
+
+  return get_t_min_() < stamp && stamp <= get_t_max_();
```
done

sdiazpier commented on nestkernel/input_device.h:
```diff
@@ -0,0 +1,122 @@
+
+} // namespace
+
+#endif // RECORDING_DEVICE_H
```
done

sdiazpier commented on nestkernel/input_device.h:
```diff
@@ -0,0 +1,122 @@
+  {
+    std::string label_;  //!< A user-defined label for symbolic device names.
+    bool time_in_steps_; //!< Flag indicating if time is recorded in steps or ms
+    Name input_from_;    //!< Array of input backends to use
```
done

sdiazpier commented on nestkernel/io_manager.cpp:
```diff
@@ -250,6 +312,33 @@ IOManager::is_valid_recording_backend( Name backend_name ) const
+  nest::InputBackend* back;
+  std::map< Name, InputBackend* >::const_iterator backend;
+  printf( "This is the backend name: %s\n", backend_name.toString().c_str() );
+  backend = input_backends_.find( backend_name );
```
done

sdiazpier commented on nestkernel/recording_backend_mpi.cpp:
```diff
@@ -0,0 +1,306 @@
+{
+  // clear vector of map
+  device_map::iterator it_device;
+  for ( it_device = devices_.begin(); it_device != devices_.end(); it_device++ )
```
done

sdiazpier commented on nestkernel/io_manager.cpp:
```diff
@@ -250,6 +312,33 @@ IOManager::is_valid_recording_backend( Name backend_name ) const
+IOManager::is_valid_input_backend( Name backend_name ) const
+{
+  std::map< Name, InputBackend* >::const_iterator backend;
+  backend = input_backends_.find( backend_name );
```
done

sdiazpier commented on nestkernel/recording_backend_mpi.cpp:
```diff
@@ -0,0 +1,306 @@
+    it_device->clear();
+  }
+  comm_map::iterator it_comm;
+  for ( it_comm = commMap_.begin(); it_comm != commMap_.end(); it_comm++ )
```
done

sdiazpier commented on nestkernel/recording_backend_mpi.cpp:
```diff
@@ -0,0 +1,306 @@
+    thread tid = device.get_thread();
+    index node_id = device.get_node_id();
+
+    device_map::value_type::iterator device_it = devices_[ tid ].find( node_id );
```
done

sdiazpier commented on nestkernel/recording_backend_mpi.cpp:
```diff
@@ -0,0 +1,306 @@
+    // add the link between MPI communicator and the device (device can share the same MPI communicator
+    std::string port_name;
+    get_port( it_device->second.second, &port_name );
+    comm_map::value_type::iterator comm_it = commMap_[ thread_id ].find( port_name );
```
done

sdiazpier commented on nestkernel/recording_backend_mpi.cpp:
```diff
@@ -0,0 +1,306 @@
+  // 2) connect the thread with MPI process it need to be connected
+  // WARNING can be a bug if it's need all the thread to be connected in MPI
+  comm_map::value_type::iterator it_comm;
+  for ( it_comm = commMap_[ thread_id ].begin(); it_comm != commMap_[ thread_id ].end(); it_comm++ )
```
done

sdiazpier commented on nestkernel/recording_backend_mpi.cpp:
```diff
@@ -0,0 +1,306 @@
+  comm_map::value_type::iterator it_comm;
+  for ( it_comm = commMap_[ thread_id ].begin(); it_comm != commMap_[ thread_id ].end(); it_comm++ )
+  {
+    printf( "Connect to %s\n", it_comm->first.data() );
```
done

sdiazpier commented on nestkernel/recording_backend_mpi.cpp:
```diff
@@ -0,0 +1,306 @@
+  // Waiting until all the receptor are ready to receive information
+  const thread thread_id = kernel().vp_manager.get_thread_id();
+  device_map::value_type::iterator it_device;
+  for ( it_device = devices_[ thread_id ].begin(); it_device != devices_[ thread_id ].end(); it_device++ )
```
done

sdiazpier commented on nestkernel/recording_backend_mpi.cpp:
```diff
@@ -0,0 +1,306 @@
+  thread thread_id = kernel().vp_manager.get_thread_id();
+  // WARNING can be a bug if all the thread to send ending connection of MPI
+  comm_map::value_type::iterator it_comm;
+  for ( it_comm = commMap_[ thread_id ].begin(); it_comm != commMap_[ thread_id ].end(); it_comm++ )
```
done

sdiazpier commented on nestkernel/recording_backend_mpi.cpp:
```diff
@@ -0,0 +1,306 @@
+  {
+    int value[ 1 ];
+    value[ 0 ] = thread_id;
+    MPI_Send( value, 1, MPI_INT, 0, 1, *it_comm->second.first );
```
done

sdiazpier commented on nestkernel/recording_backend_mpi.cpp:
```diff
@@ -0,0 +1,306 @@
+  // WARNING can be a bug if all the thread to send ending connection of MPI
+  // disconnect MPI
+  comm_map::value_type::iterator it_comm;
+  for ( it_comm = commMap_[ thread_id ].begin(); it_comm != commMap_[ thread_id ].end(); it_comm++ )
```
done

sdiazpier commented on nestkernel/recording_backend_mpi.cpp:
```diff
@@ -0,0 +1,306 @@
+  {
+    int value[ 1 ];
+    value[ 0 ] = thread_id;
+    MPI_Send( value, 1, MPI_INT, 0, 2, *it_comm->second.first );
```
done

sdiazpier commented on nestkernel/recording_backend_mpi.cpp:
```diff
@@ -0,0 +1,306 @@
+  const Time stamp = event.get_stamp();
+
+  MPI_Comm* comm;
+  device_map::value_type::iterator it_devices = devices_[ thread_id ].find( device.get_node_id() );
```
done

sdiazpier commented on nestkernel/recording_backend_mpi.cpp:
```diff
@@ -0,0 +1,306 @@
+  // clear map of device
+  commMap_[ thread_id ].clear();
+  device_map::value_type::iterator it_device;
+  for ( it_device = devices_[ thread_id ].begin(); it_device != devices_[ thread_id ].end(); it_device++ )
```
done

sdiazpier commented on nestkernel/recording_backend_mpi.cpp:
```diff
@@ -0,0 +1,306 @@
+nest::RecordingBackendMPI::get_port( const index index_node, const std::string& label, std::string* port_name )
+{
+  std::ostringstream
+    basename; // path of the file : path+label+id+.txt (file contains only one line with name of the port
```
done

sdiazpier commented on nestkernel/recording_backend_mpi.cpp:
```diff
@@ -0,0 +1,306 @@
+  thread thread_id = kernel().vp_manager.get_thread_id();
+  // get port and update the list of device
+  device_map::value_type::iterator it_device;
+  for ( it_device = devices_[ thread_id ].begin(); it_device != devices_[ thread_id ].end(); it_device++ )
```
done

sdiazpier commented on nestkernel/recording_backend_mpi.cpp:
```diff
@@ -0,0 +1,306 @@
+  }
+  else
+  {
+    // TODO take in count this case
```
done

jougs commented:
> > Would it make sense to have a common IOBackend baseclass for stimulating and recording backends? That way we could have a single registry for them in IOManager and things like initialization, parameter getting/setting could become a bit simpler and we would have to duplicate less code. This does not have to happen right away, I'd rather like to hear your thoughts on this.
>
> For my part, I don't think that there are not some advantages for the moment to have a common class between stimulating and recording backends for the moment. After removing the unnecessary function in the stimulation backend, you can see there are few functions shared between them.
I prefer to think that redesign of the IOBackend will be necessary when there are more stimulating backends. If you want to do it now, this will require some think of the type of input can be implemented in the future. I don't have the time for it.

Thank you for your assessment. As this was anyway rather meant as a point to discuss than something to implement right away, I am happy with your answer and consider this as resolved..

jougs commented:
Do you want me to re-review already, or is there more to come? I'm in any case quite impressed by the volume of recent changes ;-)

sdiazpier commented:
Dear @jougs there is more to come, I need to solve some small things still and I hope I can have this ready by the end of the weekend.

sdiazpier commented on nestkernel/input_backend_mpi.cpp:
```diff
@@ -0,0 +1,297 @@
+  }
+  char add_path[ 150 ];
+  sprintf( add_path, "/%zu.txt", index_node );
+  basename << add_path;
```
done

sdiazpier commented on nestkernel/CMakeLists.txt:
```diff
@@ -123,11 +125,22 @@ endif ()
-       )
+    ${nestkernel_sources}
+    recording_backend_arbor.h recording_backend_arbor.cpp
+    recording_backend_mpi.h recording_backend_mpi.cpp
+    stimulating_backend_mpi.h stimulating_backend_mpi.cpp
```
done

sdiazpier commented on nestkernel/recording_backend_mpi.h:
```diff
@@ -0,0 +1,131 @@
+#include <unistd.h>
+#include <mpi.h>
+
+/* BeginDocumentation
+
+Send data with MPI
+##################
+
+When a recording device sends data to the ``mpi`` backend, it sends the
+event using MPI. Events are sent with the GID and the time stamp.
+
+Communication Protocol: (value,number,type,source/destination,tag)
++++++++++++++++++++++++
+1) Connection of MPI port include in one file ( path+label+id+.txt )
+2) Send at each beginning of the run (true, 1, CXX_BOOL, 0, 0)
+3) Receive at each ending of the run (true, 1, CXX_BOOL, 0, 0)
+4) Send shape of the data of the run (shape, 1,I NT, 0, 0)
+5) Send data of the data of the run (data, shape, DOUBLE, 0, 0)
+6) Send at each ending of the run (true, 1, CXX_BOOL, 0, 1)
+7) Send at this en of the simulation (true, 1, CXX_BOOL, 0, 2)
+
+@author Lionel Kusch and Sandra Diaz
+@ingroup NESTio
+
+EndDocumentation */
```
Hopefully addressed correctly, please check.

sdiazpier commented:
I have addressed all open comments but I still need to check the set_data_from_stimulating_backend from the spike generator, there seems to be some strange condition.

sdiazpier commented:
Dear @jougs and @hakonsbm, we are ready for a next review of the code. We need to discuss some small detail of the spike generator update from backend but you could in theory already start taking a look at the current status of the PR. Hopefully we have mostly addressed all your concerns.
I also think it would be a good idea to consider having a common class for stimulating and recording backends but we can definitely address this in a follow up PR.

hakonsbm commented on models/spike_generator.h:
```diff
@@ -252,8 +237,8 @@ class spike_generator : public DeviceNode
-    Parameters_();                               //!< Sets default parameter values
-    Parameters_( const Parameters_& ) = default; //!< Recalibrate all times
+    Parameters_();                     //!< Sets default parameter values
+    Parameters_( const Parameters_& ); //!< Recalibrate all times
```
The copy constructor should remain defaulted. In C++ , implicit declaration of `operator=` has been deprecated when a user-provided copy constructor is defined.

sdiazpier replied:
Done

hakonsbm commented on nestkernel/stimulating_device.h:
```diff
@@ -88,27 +90,69 @@ namespace nest
+    Name stimulus_source_; //!< Origin of the stimulating signal.
+
+    Parameters_();
+    Parameters_( const Parameters_& );
```
The copy constructor should be defaulted here, not in the cpp file.

sdiazpier replied:
Done

hakonsbm commented on nestkernel/stimulating_device.cpp:
```diff
@@ -0,0 +1,190 @@
+{
+}
+
+nest::StimulatingDevice::Parameters_::Parameters_( const Parameters_& p ) = default;
```
The copy constructor should be defaulted in the declaration in the header file, not here.

sdiazpier replied:
Done

hakonsbm commented on nestkernel/stimulating_device.cpp:
```diff
@@ -0,0 +1,190 @@
+  , first_syn_id_( invalid_synindex ) // a new instance can have no connections
+  , backend_params_( sd.backend_params_ )
+{
+  P_ = Parameters_( sd.P_ );
```
Can this be moved up to the other initialisations, as `P_( sd.P_ )`?

sdiazpier replied:
Done

hakonsbm reviewed (changes requested):
Thanks for the changes! I have a few more suggestions.

sdiazpier commented on models/gamma_sup_generator.h:
```diff
@@ -66,6 +66,21 @@ The following parameters appear in the element's status dictionary:
                        default: 1
 ============  ======== =========================================================
 
+    ( *d )[ names::gamma_shape ] = DoubleDatum( lround( input_param[ 0 ] ) );
+    ( *d )[ names::rate ] = DoubleDatum( input_param[ 1 ] );
+    ( *d )[ names::n_proc ] = DoubleDatum( lround( input_param[ 2 ] ) );
+
+Update from stimulating backend
```
Yes sorry, must have pasted it inadvertently :(

sdiazpier commented on pynest/nest/server/hl_api_server.py:
```diff
@@ -212,7 +212,7 @@ def get_restricted_globals():
         if obj is not None and type(obj) in (list, tuple, dict, nest.NodeCollection):
             return obj[index]
-        msg = f"Error while getting restricted globals: unidentified object '{obj}'."
+        msg = "Error while getting restricted globals: unidentified object '{}'.".format(obj)
```
We changed it because f-strings are only compatible with Python 3.6 on. If support from NEST is only intended from this version of Python on, then we can change it back.

hakonsbm replied:
NEST master currently requires (although somewhat loosely) Python 3.6 or newer. Also, with #1954 it will enforce a requirement of Python 3.8 or newer.

sdiazpier commented on models/poisson_generator_ps.cpp:
```diff
@@ -284,5 +280,26 @@ nest::poisson_generator_ps::event_hook( DSSpikeEvent& e )
+
+  // if we get here, temporary contains consistent set of properties
+  P_ = ptmp;
+}
```
Done

sdiazpier commented on models/step_current_generator.cpp:
```diff
@@ -328,3 +320,46 @@ nest::step_current_generator::handle( DataLoggingRequest& e )
+        "[ sequence of the couple (time,amplitude) ] " );
+    }
+    DictionaryDatum d = DictionaryDatum( new Dictionary );
+    std::vector< double > times_ms;
+    std::vector< double > amplitudes_pA;
```
Done

sdiazpier commented on models/step_rate_generator.cpp:
```diff
@@ -340,3 +333,46 @@ nest::step_rate_generator::handle( DataLoggingRequest& e )
+        "[ sequence of the couple (time,rate) ] " );
+    }
+    DictionaryDatum d = DictionaryDatum( new Dictionary );
+    std::vector< double > times_ms;
+    std::vector< double > amplitudes_Hz;
```
Done

sdiazpier commented on models/step_rate_generator.h:
```diff
@@ -263,12 +252,24 @@ step_rate_generator::set_status( const DictionaryDatum& d )
   P_ = ptmp;
 }
 
+//! Allow multimeter to connect to local instances
```
No, it should be the step rate generator. Corrected now.

sdiazpier commented on nestkernel/recording_backend_mpi.h:
```diff
@@ -0,0 +1,155 @@
+  bool enrolled_;
+  bool prepared_;
+  /**
+   * Buffer for saving events before to send it.
+   * The buffer sa 3 dimensions : thread_id, MPI_communicator_index and number of events
```
Fixed

sdiazpier commented on nestkernel/recording_backend_mpi.h:
```diff
@@ -0,0 +1,155 @@
+  typedef std::vector< std::map< index, std::tuple< int, MPI_Comm*, const RecordingDevice* > > > device_map;
+  device_map devices_;
+  /**
+   * A map of MPI communicator use by the master thread for the MPI communication.
+   * This map contains also the number of the device by MPI communicator.
```
Fixed

sdiazpier commented on nestkernel/stimulating_backend_mpi.cpp:
```diff
@@ -0,0 +1,423 @@
+    else
+    {
+      // create a new MPI communicator for the MPI process. Only the master thread is using MPI function.
+      // The cause it's the management of the thread is MPI_THREAD_FUNNELED (mpi_manager 119).
```
Tried to clarify this, please confirm if this is better now

hakonsbm replied:
Yes, it's better now. Thanks!

sdiazpier commented on nestkernel/stimulating_backend_mpi.h:
```diff
@@ -0,0 +1,149 @@
+  using device_map = std::vector< std::map< index, std::pair< const MPI_Comm*, StimulatingDevice* > > >;
+  device_map devices_;
+  /**
+   * A map of MPI communicator use by the master thread for the MPI communication.
+   * This map contains also the number of the device by MPI communicator.
```
Fixed

sdiazpier commented on nestkernel/stimulating_device.h:
```diff
@@ -25,14 +25,16 @@
-class DelayedRateConnectionEvent;
+// Includes from libnestutil:
+#include "compose.hpp"
+#include <string>
```
Done

sdiazpier commented on nestkernel/target_table_devices.h:
```diff
@@ -201,6 +201,11 @@ class TargetTableDevices
     const index lcid );
+
+  /**
+   * Checks if the device has any connection in this thread
```
Done

sdiazpier commented on nestkernel/target_table_devices.h:
```diff
@@ -201,6 +201,11 @@ class TargetTableDevices
+  /**
+   * Checks if the device has any connection in this thread
+   */
+  bool get_devices_connected( thread tid, index lcid ) const;
```
Done

sdiazpier commented on nestkernel/stimulating_device.h:
```diff
@@ -88,27 +90,69 @@ namespace nest
+    throw KernelException( "WORNG TYPE" );
+  };
+  const std::string& get_label() const;
+  virtual void set_data_from_stimulating_backend( std::vector< double > ){};
```
Done

sdiazpier commented on nestkernel/target_table_devices.h:
```diff
@@ -241,6 +246,25 @@ TargetTableDevices::send_from_device( const thread tid,
+inline bool
+TargetTableDevices::get_devices_connected( const thread tid, const index lcid ) const
+{
+  for ( auto& it_synapse : target_from_devices_[ tid ][ lcid ] )
```
Done

sdiazpier commented:
Dear @hakonsbm I have now hopefully addressed all your comments and suggestions, thank you very much for the detailed review. Please let me know if you have still some open things to look into.

hakonsbm commented on nestkernel/recording_backend_mpi.h:
```diff
@@ -0,0 +1,156 @@
+  device_map devices_;
+  /**
+   * A map of MPI communicators used by the master thread for the MPI communication.
+   * This map contains also the number devices linked to each MPI communicator.
```
I think this is still a bit too vague. I suggest to explicitly list what the tuple contains:
```suggestion
   * The values of the map are tuples containing the index of the MPI communicator, the MPI communicator itself,
   * and the number of devices linked to that MPI communicator.
```
Please verify that I have correctly guessed the contents of the tuple. :)

lionelkusch replied:
it's correct

hakonsbm commented on nestkernel/stimulating_backend.h:
```diff
@@ -0,0 +1,179 @@
+  virtual void enroll( StimulatingDevice&, const DictionaryDatum& ){};
+
+  /**
+   * Disenroll an `StimulatingDevice` from the `StimulatingBackend`.
```
```suggestion
   * Disenroll a `StimulatingDevice` from the `StimulatingBackend`.
```

hakonsbm commented on nestkernel/stimulating_backend_mpi.h:
```diff
@@ -0,0 +1,149 @@
+  /**
+   * A map for the enrolled devices. We have a vector with one map per local
+   * thread. The map associates the node ID of a device on a given thread
+   * with it's device. Only the master thread has a valid MPI communicator pointer.
```
```suggestion
   * with its device. Only the master thread has a valid MPI communicator pointer.
```

hakonsbm reviewed (changes requested):
I just have a couple more suggestions, then I'm happy.

jougs commented:
@sdiazpier, @lionelkusch: I'll re-review as soon as @hakonsbm's recent comments have been addressed.

hakonsbm reviewed (approved):
Thanks!

jougs commented on nestkernel/stimulating_backend.h:
```diff
@@ -0,0 +1,179 @@
+
+namespace nest
+{
+
```
Please add some developer-level documentation about the class describing the basic operations for `StimulatingBackend`s. See the corresponding [documentation of `RecordingBackend`](https://github.com/nest/nest-simulator/blob/master/nestkernel/recording_backend.h#L39-L67) for what I mean.

lionelkusch replied:
I add the document based on the corresponding documentation.

jougs commented on nestkernel/stimulating_backend.h:
```diff
@@ -0,0 +1,179 @@
+
+} // namespace
+
+#endif // STIMULATING_BACKEND_H
```
All Doxygen comments are _still_ (more or less) just copies of the ones from `recording_backend.h`. Even the term `recording` is not replaced everywhere, which makes me doubt that the documentation is correct or appropriate like it is.

As an example, the documentation still talks about the non-existent function `set_value_names()` and the two static members don't make any sense to me in this class. Please update the Doxygen comments to reflect reality, and describe the interface of this class instead of just replacing terms. Thanks!

lionelkusch replied:
The documentation of the functions is updated for the stimulated backend.

jougs commented on nestkernel/stimulating_backend.cpp:
```diff
@@ -0,0 +1,26 @@
+#include "stimulating_backend.h"
+
+const std::vector< Name > nest::StimulatingBackend::NO_DOUBLE_VALUE_NAMES;
+const std::vector< Name > nest::StimulatingBackend::NO_LONG_VALUE_NAMES;
```
I don't think this file is actually needed. See my comment on the corresponding header file.

lionelkusch replied:
This variable and function are deleted because there are not used.

jougs commented on nestkernel/stimulating_backend_mpi.h:
```diff
@@ -0,0 +1,149 @@
+#include <unistd.h>
+#include <mpi.h>
+
+/* BeginDocumentation
+
+Read data from MPI
+##################
+
+When an stimulating device is to be updated at the beginning of each step run, the 'mpi' backend
+communicates with an external source via MPI to get the update.
+
+Communication Protocol: (value,number,type,source/destination,tag)
++++++++++++++++++++++++
+1) Connection of MPI port include in one file ( path+label+id+.txt )
+2) Send at each beginning run of Nest (true, 1, CXX_BOOL, 0, 0)
+3) Send the id of the device to update (id_device, 1, INT, 0, 0)
+3) Receive shape of the data (shape, 1, INT, 0, 0)
+4) Receive the data for updating the device (data, shape, DOUBLE, 0, 0)
+5) Send at each ending of the run (true, 1, CXX_BOOL, 0, 1)
+6) Send at this en of the simulation (true, 1, CXX_BOOL, 0, 2)
+
+
+@author Lionel Kusch and Sandra Diaz
+@ingroup NESTio
+
+EndDocumentation */
```
Please make this into a user documentation block as it is for example in [`recording_backend_ascii.h`](https://github.com/nest/nest-simulator/blob/master/nestkernel/recording_backend_ascii.h#L31-L139) so it can be properly picked up by our Sphinx pipeline.

Also, please add some information to the [Guide for stimulating the network](https://github.com/nest/nest-simulator/blob/master/doc/userdoc/guides/stimulating_the_network.rst) that explains how to use stimulation backends from a user perspective.

jougs reviewed (changes requested):
Thanks for your updates. I am mostly happy with the code now and all my more substantial comments are about the state of the developer- and user-level documentation.

I've also added some stereotypical suggestions that you can probably just batch commit.

Please also merge master and resolve any remaining conflicts.

lionelkusch commented on nestkernel/io_manager.cpp:
```diff
@@ -277,7 +327,41 @@ IOManager::enroll_recorder( Name backend_name, const RecordingDevice& device, co
+nest::IOManager::enroll_stimulator( const Name& backend_name, StimulatingDevice& device, const DictionaryDatum& params )
+{
+
+  if ( not is_valid_stimulating_backend( backend_name ) and backend_name.toString() != "internal" )
```
no, we can remove internal.

lionelkusch commented on nestkernel/io_manager.cpp:
```diff
@@ -277,7 +327,41 @@ IOManager::enroll_recorder( Name backend_name, const RecordingDevice& device, co
+    {
+      if ( it.first == backend_name )
+      {
+        if ( backend_name.toString() == "mpi" )
```
This if it's not need it. I think that we forget to remove it.

lionelkusch commented:
Hi @jougs,
I did most of the modification. However, I have a few questions. 

In the "Guide for stimulating the network", do you think the last line is necessary?
 `.. doxygengroup:: generator
   :content-only:` 
   
I would like to add an image like the following somewhere in the documentation :
  
[communication_Nest.pdf](https://github.com/nest/nest-simulator/files/6451003/communication_Nest.pdf)

Do you think that I should create a common documentation page for the mpi backends (recording and stimulating) or 2 different pages? 

I also have this example of communication based on state diagram :
[Nest_IO.pdf](https://github.com/nest/nest-simulator/files/6451022/Nest_IO.pdf)
[state_Nest.pdf](https://github.com/nest/nest-simulator/files/6451023/state_Nest.pdf)

Do you think I should add them somewhere? 

I also update for the recording backend. Can you check if I don't miss something?


@Sandra : Can you have a check of my modifications and merge master?

Have a good day,
Lionel

jougs commented on nestkernel/stimulating_backend_mpi.h:
```diff
@@ -0,0 +1,191 @@
+updates stimulating device just before each run. This is useful for
+co-simulation or for receiving data from an external software.
+
+The creation of this MPI communication is based on the function 
+'MPI_Comm_connect' and 'MPI_Comm_disconnect'. The port name is 
+getting from a file for each device with this backend.
+The name of the file need to be named according to the following 
+pattern:
```
```suggestion
The creation of this MPI communication is based on the functions
'MPI_Comm_connect' and 'MPI_Comm_disconnect'. The port name is 
read from a file for each device with this backend.
The name of the file needs to be specified according to the following 
pattern:
```

jougs commented on nestkernel/stimulating_backend_mpi.h:
```diff
@@ -0,0 +1,191 @@
+
+   data_path/label/id_device.txt
+
+The properties ``data_path`` and ``data_prefix`` are global kernel properties.
+This path is only used during the ``Prepare`` call. There is not
+functionality for changing connection during the ``Run``.
+If you want to change the connections during the simulation, you need first 
+to close the previous one by using calling ``Cleanup`` of the simulator.
```
```suggestion
The ``data_path`` is a global kernel property, while `label` is a
property of the device in question and `id_device` its node ID. 
This path can only be set outside of a `Run` contexts (i.e. not
after ``Prepare()` has been called, but ``Cleanup()`` has not).
```

sdiazpier replied:
Did some small corrections and added your changes @jougs

jougs commented on nestkernel/stimulating_backend_mpi.h:
```diff
@@ -0,0 +1,191 @@
+   Shared file with ports with the format path+label+id+.txt
+   Where path = kernel().io_manager.get_data_path() and the
+   label and id are the ones provided for the device.
+
+@author Lionel Kusch and Sandra Diaz
+@ingroup NESTio
+
```
```suggestion
```

jougs reviewed (changes requested):
Many thanks for your updates! I still have a couple of suggestions.

jougs commented:
> In the "Guide for stimulating the network", do you think the last line is necessary?
> `.. doxygengroup:: generator :content-only:`

This can go away. The page should probably start with a general introduction of how stimulation devices work without any stimulation backend present. It could then link to the list of available backends.

> I would like to add an image like the following somewhere in the documentation :
> 
> [communication_Nest.pdf](https://github.com/nest/nest-simulator/files/6451003/communication_Nest.pdf)
> 
> Do you think that I should create a common documentation page for the mpi backends (recording and stimulating) or 2 different pages?

The documentation of the two backends is good after my latest comments have been addressed. They should just be linked in the corresponding guides for [Recording](https://github.com/nest/nest-simulator/blob/master/doc/userdoc/guides/recording_from_simulations.rst) from and [Stimulating](https://github.com/nest/nest-simulator/blob/master/doc/userdoc/guides/stimulating_the_network.rst) the network.

In general, I tend to think more illustrations are better than less, but the one you're pointing at is quite complex. Maybe it should rather be linked from the developer level (i.e., Doxygen) documentation. The same is true for these:

> [Nest_IO.pdf](https://github.com/nest/nest-simulator/files/6451022/Nest_IO.pdf)
> [state_Nest.pdf](https://github.com/nest/nest-simulator/files/6451023/state_Nest.pdf)

Please also fix the conflicts.

After these points have been addressed, I think we're good to go!

sdiazpier commented on nestkernel/stimulating_backend.h:
```diff
@@ -0,0 +1,208 @@
+   *
+   * @ingroup NESTio
+   */
+  virtual void post_step_hook() = 0;
```
I removed the function and added some comments about it. Maybe worth taking a look if the comments are in the right place.

jougs commented on nestkernel/stimulating_backend_mpi.h:
```diff
@@ -0,0 +1,185 @@
+
+   data_path/label/id_device.txt
+
+The ``data_path`` and ``data_prefix`` are global kernel properties,
```
I don't see `data_prefix` used anywhere in the pattern. Either add it to the pattern or remove it from the text (as I suggested earlier). The way it is right now, it is confusing to have it explained.

The same changes need to be applied to the docstring of `RecordingBackendMPI`.

sdiazpier replied:
The data_prefix is actually used, I modified the pattern to make it evident, sorry for that.

jougs reviewed (commented):
Thanks for your changes! Here are three more comments.

Pleas consider using the "batch commit" feature in the "Files Changed" pane in the future. That way you could bundle multiple suggestions into a single commit and also reduce the number of notification mails sent by GitHub. Thanks!

sdiazpier commented:
Dear @jougs about the diagrams, communication_Nest.pdf, Nest_IO.pdf and state_Nest.pdf, where would it be good to store them so we can reference from the developer documentation?
I hope all the other comments are now solved :)

jougs commented:
@sdiazpier, many thanks. All my suggestions are addressed!

Regarding the diagrams: @jessica-mitchell, what would be an appropriate location for such static content of the developer documentation? I tend to think that `doc/devdoc/static` would look nice alongside the already existing `doc/userdoc/static`. In any case, the Doxyfiles would have to be configured (and tested) accordingly, as this is the first time that we have custom files to be included in the developer documentation.

lionelkusch commented:
Dear @jougs and @jessica-mitchell,
These are the image in png format to add to the Nest documentation :
![communication_Nest](https://user-images.githubusercontent.com/17182418/118612683-5012a300-b7be-11eb-866f-6d857f967575.png)
![state_Nest](https://user-images.githubusercontent.com/17182418/118612685-50ab3980-b7be-11eb-931e-b1ff938aa983.png)
![Nest_IO](https://user-images.githubusercontent.com/17182418/118612688-5143d000-b7be-11eb-9c61-3b1e3aa8d58f.png)

jougs commented:
I would appreciate if the figures were added also in a vector format or any source format they were generated from. The rationale for this is that we might need to tweak the figures in the future. I suggest to add them to `doc/devdoc/static` for the time being.

I would be OK with approving this without more extensive developer documentation, but would like to see at least some basic information about the idea on the user level stimulation guide. Also the links to the MPI backend documentation in the recording and stimulation guides are needed, so they can actually be found.

After reviewing #1929, I think we also need a section on the availability of the Stimulating Backend infrastructure in the NEST 2->3 guide. @jessica-mitchell, can you please give a hint on where this should best go?

jessica-mitchell commented:
@jougs I'd suggest if it's a user level document to highlight the new feature, it should go here: `doc/userdoc/guides/nest2_to_nest3/nest3_features/`_or_ put it in `doc/userdoc/guides/` if it's much more in-depth and detailed.  _And_ include a link here `doc/userdoc/guides/nest2_to_nest3/nest3_features/index.rst` which also contains links to:

-  New functionality for node handles (neurons and devices)
-  New functionality for handling connections (synapses)
-  Parametrization
-  Simpler handling of random number generators
-  Recording from simulations
-  NEST Server


Also the _devdoc/static_ sounds fine to me too

sdiazpier commented:
Dear @lionelkusch thanks so much for adding the figures and documentation. Dear @jougs I think we are ready for your final checks. Hopefully everything is now covered.

jougs commented:
@lionelkusch, @sdiazpier: Many thanks!

I have just sent a PR towards your branch, in which I mostly reworked the structure of the user level documentation. I think you don't have to look into it in detail, as I already discussed most changes with @jessica-mitchell and @terhorstd. Once that's merged, I'll merge this one.

sdiazpier commented:
Dear @jougs and everyone involved in the PR, I just merged your changes to our branch. Thanks so much for the revised documentation.

jougs reviewed (approved):
Also many thanks from my side for all the hard work!
